### ISSUE TITLE
Display parallel test execution through console

### ISSUE THREAD

**Issue by eriwen** [gradle/gradle#1612]
### Expected Behavior
Gradle can be configured to fork multiple test workers, and these test workers should show up independently on the console.

### Current Behavior
All test work is grouped into one progress operation and doesn't properly make user aware of parallel work under the hood

### Context
Considered to be one of the biggest missing things from the new console from internal discussions. This story needs a related design story before it can be prioritized.

**Comment by bmuschko** [comment]
## Findings from spike

While spiking this feature I considered two different options for rendering test worker information in the work-in-progress section.

### Option 1

Rendering test worker information as part of the existing test task work-in-progress line.

#### Mockups

Render overall test executions, show number of tests executed by workers:

```
> :other:compileJava
> :fooBarBazQuux:test [134 completed, 2 skipped] > (Worker 1) 35 completed, (Worker 2) 99 completed, (Worker 3) 2 skipped`
> IDLE
```

Render overall test executions, show number of workers and the test classes they work on:

```
> IDLE
> :fooBarBazQuux:test [134 completed, 2 skipped] > (3 workers running) FooBarIntegrationTest, BazQuuxIntegrationTest, OtherTest
> :other:compileJava
````

#### Tradeoffs

- **Pros:**
    - Condensed and focused view on test worker information
    - WIP line number doesn't need change and therefore doesn't draw away attention.
    - Easier to implement
- **Cons:**
    - Less real estate to show relevant information
    - Information might cut off at the end of line

### Option 2

Rendering test worker information as new work-in-progress lines underneath the test task line.

#### Mockups

Render overall test executions on dedicated WIP line, each worker is displayed on its own line:

```
> IDLE
> :plugin:functionalTest > 134 tests completed, 2 skipped
      > Worker 17 > 122 tests completed, 2 skipped
      > Worker 19 > 2 tests completed
> :other:compileJava
> :fooBarBazQuux:test > 3 tests completed
     > Worker 25 > 3 tests completed
> IDLE
```

Also renders test class for each worker:

```
> IDLE
> :plugin:functionalTest > 134 tests completed, 2 skipped
      > Worker 17 > 122 tests completed, 2 skipped > FooBarIntegrationTest
      > Worker 19 > 2 tests completed > BazQuuxIntegrationTest
> :other:compileJava
> :fooBarBazQuux:test > 3 tests completed
     > Worker 25 > 3 tests completed > OtherTest
> IDLE
```

#### Tradeoffs

- **Pros:**
    - Test worker information can be pretty detailed
    - The top level WIP line would look as before -> renders total number of tests executed
    - The view on a test execution would rather resemble a tree than a flattened list
- **Cons:**
    - Bigger implementation effort
        - WIP are needs to grow dynamically
        - Rendering of output needs to be changes to avoid stray output
        - WIP area should probably also need to shrink
    - WIP lines can become very long depending on the configured number of `maxParallelForks`

## Proposed solution

Go with option 1 as it is less effort to implement but adds immediated benefit. Upon user feedback option 2 can be considered.

### User visible changes

The existing WIP line for tests will add information about the test workers.

### Implementation

- Extend existing implementation in `TestCountLogger` to collect information about workers.
- Traverse `TestDescriptor` hierarchy, find `WorkerTestClassProcessor.WorkerTestSuiteDescriptor` instance to retrieve information about test worker.
- Traverse `TestDescriptor` hierarchy, find `DefaultTestDescriptor` instance to retrieve information about executed test class.
- Collect overall test execution metrics as before.
- Additionally collect test execution metrics for test workers in a `LinkedHashMap`.
- Hook into `beforeTest`, `afterTest`, `afterSuite` to manage test execution metrics.
- Render the overall summary as part of the existing `summary()` method.

### Test coverage

- Overall test execution metrics are always rendered independent of number of test workers.
- Test workers are rendered in the WIP line when in progress.
- Test workers are removed from WIP line when finished.
- Test workers potentially grow with the configured number of `maxParallelForks`.
- Test worker information is rendered in the order of creation.
- A test workers that executes different test classes in order (e.g. when `forkEvery` is configured) reflect that in the output.
- Information in the WIP line that exceeds the maximum displayable characters is cut off.

### Open issues

- Traversing the `TestDescriptor` hierarchy for every event might degrade performance.

**Comment by eskatos** [comment]
Hey Ben,

In a not so distant future we'll want to do about the same for tasks using the `WorkerExecutor` API.

Here are some mockups of a fictitious `:doMyStuff` task that would use the `WorkerExecutor` API to do stuff in parallel. Each unit of work has a display/progress name. The main difference would be that the overall task has no display/progress name.

### Option 1

```
> :other:compileJava
> :some:doMyStuff > (Worker 11) Parsing foo, (Worker 17) Munging bar, (Worker 23) Generating bazar 
> IDLE
```
or
```
> :other:compileJava
> :some:doMyStuff > (3 workers running) Parsing foo, Munging bar, Generating bazar 
> IDLE
```

I'm not sure this option convey parallelism very well as a single line represent multiple things happening in parallel.

### Option 2

```
> IDLE
> :some:doMyStuff
      > Worker 11 > Parsing foo
      > Worker 17 > Munging bar
      > Worker 23 > Generating bazar
> :other:compileJava
> :fooBarBazQuux:test > 3 tests completed
> IDLE
```

This uses 4 lines where 3 parallel executions are happening.

### Suggestion

Just a suggestion, what about using lines to convey parallelism, repeat the task path and remove the not so useful worker names?

For test execution this would look like:

```
> IDLE
> :plugin:functionalTest > 122 tests completed, 2 skipped
> :plugin:functionalTest > 2 tests completed
> :other:compileJava
> :fooBarBazQuux:test > 3 tests completed
> IDLE
```

The main con here would be that you don't get the overall test count. But this may be acceptable.

And for a task using the `WorkerExecutor` API:

```
> IDLE
> :some:doMyStuff > Parsing foo
> :some:doMyStuff > Munging bar
> :some:doMyStuff > Generating bazar
> :other:compileJava
> :fooBarBazQuux:test > 3 tests completed
> IDLE
```

One thing could be seen as weird here, a task path is repeated. But I think it reflects reality: a single task is being executed by several workers.

WDYT?

**Comment by bmuschko** [comment]
> In a not so distant future we'll want to do about the same for tasks using the `WorkerExecutor` API.

Thanks for the heads up. Can you describe a concrete use case for a task using multiple workers?

Would this information be valuable to the user? I am wondering if you could just add a similar mechanism like `Test.testLogging` for rendering it on demand. One of the goals of the new console is to show _less_ information than before.

> I'm not sure this option convey parallelism very well as a single line represent multiple things happening in parallel.

I think it depends on what we want to communicate with the information and how to make it parseable for the user. Especially when thinking about test execution, the information might only pop up and leave the screen for a split second as unit tests can be executed very quickly. I feel that it might be hard to follow for a user when lines change constantly. At least at the moment the lines also do not seem to be ordered so the information can show up in any of WIP lines. That's easily fixable I guess though.

> One thing could be seen as weird here, a task path is repeated. But I think it reflects reality: a single task is being executed by several workers.

Personally, I'd rather see a tree representation than a somewhat repeated representation. It's something I thought about in the beginning as well but didn't consider shortly thereafter. It also raises the question for me again what a line _really_ represents. I _think_ it represents the number of processors on the machine (e.g. CPUs or cores). It's really the case for test forks which is independent of this number as it is configured with `maxParallelForks`.

I wouldn't want to rule out this option but it would be my least preferred option.

> The main con here would be that you don't get the overall test count. But this may be acceptable.

I think users will want to know the overall test count number as the main indication about how long things will still take. They care less about the workers from my perspective. Losing this information might not be received well. But I am speculating here.

**Comment by eskatos** [comment]
Well, `WorkerExecutor` is general purpose to both parallelize and isolate unit of work. That's pretty close to test execution. We did not start to dogfood yet but the upcoming Android plugin is using it. Build operations are already emitted around each unit of work.

As to what lines represent, they are `workers` to me, as in `--max-workers`.
So it fits with test executors and `WorkerExecutor` very well.
As a user I expect that if there are 3 lines showing progress, then 3 workers are in use. If I did set `--max-workers` to 8 then this also tells me that my build is not parallelized enough and I need to divide work. The build scan timeline tells you this too, but this is instant feedback.

I agree about the total number of executed tests.
A tree-like structure sounds good to me as long as it doesn't get too complex.

**Comment by bmuschko** [comment]
`--max-workers` and `maxParallelForks` are orthogonal though. The might be a "worker" in concept but it would definitely have to change the number of WIP lines.

@eriwen Do you have anything to say about @eskatos feedback?

**Comment by eskatos** [comment]
Test execution honors `maxWorkers` and `maxParallelForks` is capped at `maxWorkers`.

It means that if you e.g. set `maxWorkers` to 8 and you have 3 test tasks each with `maxParallelForks` set to 12:
- `maxParallelForks` is forced to 8
- there won't ever be more than 8 parallel test execution even if the 3 test tasks execute in parallel

**Comment by bmuschko** [comment]
Thanks, @eskatos. That's important information to have. So far my assumption was that they are independent of each other. That fact somewhat does change my view on a proposed solution. They only reservation I have is the fact that a user wouldn't be able see the overall count of executed tests.

**Comment by eriwen** [comment]
I'm a bit conflicted on what to do here. 

On one hand, separate lines would consistently represent a parallel process, and that is very nice. There is hidden complexity here to group test workers because we currently map progress operations to lines so they don't jump around as new ones start and complete — so doing what's suggested by @eskatos would be a non-trivial amount of additional work. I weakly share @bmuschko's concern that we lose an overall summary unless we can come up with a better design.

On the other hand, what's suggested by @bmuschko is straightforward to implement and incrementally gets the user better information at the cost of breaking the "line per process" paradigm. It's not a future-proof solution, but it gets us incrementally better. I vote weakly for this option, but think that a form of @eskatos's solution is medium/longer term where we want to be.

I encourage you both to optionally spend a bit more time to see if we can come up with something even better. Be creative. We can fall back to what we know what will work after exploring a bit.

**Comment by eskatos** [comment]
Some thoughts about option 2 above.

Let's say I have `maxWorkers = 8` and that at some point, before running any tests, I had 8 tasks running in parallel, just because.

I would expect the output to be as shown in the post ;
```
> IDLE
> :some:doMyStuff
      > Worker 11 > Parsing foo
      > Worker 17 > Munging bar
      > Worker 23 > Generating bazar
> :other:compileJava
> :fooBarBazQuux:test > 3 tests completed
> IDLE
```
that is, 8 lines.

But I suspect that with the current implementation that only grows the number of shown lines it could end up like that instead  ;
```
> IDLE
> :some:doMyStuff
      > Worker 11 > Parsing foo
      > Worker 17 > Munging bar
      > Worker 23 > Generating bazar
> :other:compileJava
> :fooBarBazQuux:test > 3 tests completed
> IDLE
> IDLE
> IDLE
> IDLE
```
that is, 11 lines (8 workers + 3 test executions) which doesn't make sense to me.
My assumptions about the current implementation might be completely wrong here, please correct me.

**Comment by eskatos** [comment]
Yeah, missing the overall test execution progress is not nice.

We could drop the "progress per worker" display and only show the overall progress as we did before but simply take the parallelism into account:
```
> IDLE
> :plugin:functionalTest > 127 completed, 2 skipped
> :plugin:functionalTest > 127 completed, 2 skipped
> :other:compileJava
> :fooBarBazQuux:test > 3 completed
> IDLE
```
This just duplicates the `:plugin:functionalTest` line because it's actually running in //, but doesn't detail progress of the nested parallelization.

We could then look at adding the "progress per worker", e.g.:
```
> IDLE
> :plugin:functionalTest > 124 completed, 2 skipped > 122 completed, 2 skipped
> :plugin:functionalTest > 124 completed, 2 skipped > 2 completed
> :other:compileJava
> :fooBarBazQuux:test > 3 completed
> IDLE
```

For tests execution it may looks a bit like repeating. But it could work of all parallel executions.
WDYT?

**Comment by eriwen** [comment]
This might be workable, though I would say that "progress per worker" is more noise than it's worth. What we could do is say `(worker 1)` somewhere in there to make it clear that it's not actually a bug that users are seeing multiple lines.

Grouping these lines would be nice but not critical; it would require a different strategy for attaching labels to their assigned lines.

**Comment by eriwen** [comment]
Ok, after trying out @bmuschko's branch, I really like showing the test class as well. It really gives a sense of slow tests and "feels" more informative. Having it on separate lines would allow this as well and add to the parallelism notion.

**Comment by bmuschko** [comment]
> that is, 11 lines (8 workers + 3 test executions) which doesn't make sense to me.
My assumptions about the current implementation might be completely wrong here, please correct me.

The description was based on my assumption that workers and JVM test forks are orthogonal. I just didn't update the implementation description yet after you corrected me about the usage of workers. So the console won't have to grow and shrink. Do you know if the test task itself is considered a "worker" by itself and therefore would take away from the overall number of workers?

> We could drop the "progress per worker" display and only show the overall progress as we did before but simply take the parallelism into account.

I am personally not a fan of duplicating the total test count for every worker. It feels like a lot of noise. Losing the total test count completely isn't an option IMHO. Here's why I think we should display the overall test count at any time:

1. It gives an indication of progress in the context of all test cases executed by a test task.
2. A user might decide to cancel test execution if at least one test failed in the process.

> On the other hand, what's suggested by @bmuschko is straightforward to implement and incrementally gets the user better information at the cost of breaking the "line per process" paradigm. It's not a future-proof solution, but it gets us incrementally better. I vote weakly for this option, but think that a form of @eskatos's solution is medium/longer term where we want to be.

I am all for incremental progress. Releasing with the one line option might actually trigger a discussion in the community about how to show detailed information. We could even ask for feedback. Maybe also present people with some options at the Summit and see what they think.

> I encourage you both to optionally spend a bit more time to see if we can come up with something even better. Be creative. We can fall back to what we know what will work after exploring a bit.

Some other thoughts that crossed my mind that might or might not work at all:

- Being able to render additional information for a line similar by hovering over it with the mouse cursor. I have no idea if a console even allows for such functionality.
- Some new area in the console than can render more detailed information though I feel like it would feel disjunct from the information it belongs to.

Overall I still feel that a tree view is the correct representation for this kind of data if we wanted to render information on multiple lines. Other benefits: 
- It's visually easy to identify what belongs together by indenting the output.
- It's easier to preserve the order in the rendered output.
- It provides more real estate than a single line solution and just a little less than a full line.
- We can clearly indicate the worker by some identifier.

I further looked a bit at other build tools and how they display parallel execution of tests. So far I haven't found a single tool that renders parallel test execution in real-time. We are likely pioneering here.

**Comment by adammurdoch** [comment]
Test execution should be treated the same way as parallel resolution, as it's _exactly_ the same problem. That might mean changing how parallel resolution is presented, or it might mean presenting the tests in the same way as the parallel downloads currently are. We should not implement multiple ways to present the same kind of information.

**Comment by adammurdoch** [comment]
I would go with option 3 as a step towards option 2: present test execution as a sibling of the task, just like we do for artifact downloads that happen as part of a task. This will be simpler to implement than either of the other proposed options, and is far superior to option 1 in presenting the work that is happening in parallel, presenting all such work in a consistent way. Arguably it's better than option 2 as well for clarity.

**Comment by eskatos** [comment]
@bmuschko 
> Do you know if the test task itself is considered a "worker" by itself and therefore would take away from the overall number of workers?

Nope, the test task wait for it's child workers to finish, it doesn't do anything in the mean time. In that case, one child worker borrow its parent worker-lease.

> Losing the total test count completely isn't an option IMHO.

Agreed

**Comment by bmuschko** [comment]
@adammurdoch I am assuming that option 3 would look like this:

1. We have one line that represents the task execution including the overall count.
2. We have one line per forked test JVM/test worker that could render additional information e.g. the test class name.

**Example:**

```
> IDLE
> :plugin:test > 203 tests completed
> IDLE
> :plugin:test > Executing test org.gradle.api.ProjectTest
> :plugin:test > Executing test org.gradle.api.OtherTest
> IDLE
```

I played around with this a bit and was struggling with keeping the build operation running and displayed for the whole span of the execution of a test class. My approach was to use run the build operation in `TestClassProcessor.processTestClass` to begin with knowing that it would only be rendered in the console for time the processing is starting. I tried moving the build operation to `TestWorker.execute` which seems to be the right place at first glance but got the message `Some child operations have not yet completed.`. Didn't have a chance to dig deeper though.

Can you give me some more information on how we'd tackle the implementation from your point of view?
